cmd/routed-eni-cni-plugin/driver/driver.go

```diff
@@ -267,354 +271,358 @@ func WaitForAddressesToBeStable(ifName string, timeout time.Duration) error {
+		containerAddr.String(), hostVeth.Attrs().Name, "main")
 
-	log.Debugf("Successfully set host route to be %s/0", route.Dst.IP.String())
-
-	// 6. Add ip rules for the pod.
-	vlanRule := os.netLink.NewRule()
-	vlanRule.Table = vlanTableID
-	vlanRule.Priority = vlanRulePriority
-	vlanRule.IifName = vlanLink.Name
-	err = os.netLink.RuleAdd(vlanRule)
-	if err != nil && !isRuleExistsError(err) {
-		return errors.Wrapf(err, "SetupPodENINetwork: unable to add ip rule for vlan link %s ", vlanLink.Name)
+	toContainerRule := n.netLink.NewRule()
+	toContainerRule.Dst = containerAddr
+	toContainerRule.Priority = toContainerRulePriority
+	toContainerRule.Table = unix.RT_TABLE_MAIN
+	if err := n.netLink.RuleAdd(toContainerRule); err != nil && !isRuleExistsError(err) {
+		return errors.Wrapf(err, "failed to setup toContainer rule, containerAddr=%s, rtTable=%v", containerAddr.String(), "main")
 	}
 
-	vlanRule.IifName = hostVeth.Attrs().Name
-	err = os.netLink.RuleAdd(vlanRule)
-	if err != nil && !isRuleExistsError(err) {
-		return errors.Wrapf(err, "SetupPodENINetwork: unable to add ip rule for host veth %s", hostVethName)
+	log.Debugf("Successfully setup toContainer rule, containerAddr=%s, rtTable=%v", containerAddr.String(), "main")
+
+	if rtTable != unix.RT_TABLE_MAIN {
+		fromContainerRule := n.netLink.NewRule()
+		fromContainerRule.Src = containerAddr
+		fromContainerRule.Priority = fromContainerRulePriority
+		fromContainerRule.Table = rtTable
+		if err := n.netLink.RuleAdd(fromContainerRule); err != nil && !isRuleExistsError(err) {
+			return errors.Wrapf(err, "failed to setup fromContainer rule, containerAddr=%s, rtTable=%v", containerAddr.String(), rtTable)
+		}
+		log.Debugf("Successfully setup fromContainer rule, containerAddr=%s, rtTable=%v", containerAddr.String(), rtTable)
 	}
+
 	return nil
 }
 
-// buildRoutesForVlan builds routes required for the vlan link.
-func buildRoutesForVlan(vlanTableID int, vlanIndex int, gw net.IP) []netlink.Route {
-	return []netlink.Route{
-		// Add a direct link route for the pod vlan link only.
-		{
-			LinkIndex: vlanIndex,
-			Dst:       &net.IPNet{IP: gw, Mask: net.CIDRMask(32, 32)},
-			Scope:     netlink.SCOPE_LINK,
-			Table:     vlanTableID,
-		},
-		{
-			LinkIndex: vlanIndex,
-			Dst:       &net.IPNet{IP: net.IPv4zero, Mask: net.CIDRMask(0, 32)},
-			Scope:     netlink.SCOPE_UNIVERSE,
-			Gw:        gw,
-			Table:     vlanTableID,
-		},
+func (n *linuxNetwork) teardownIPBasedContainerRouteRules(containerAddr *net.IPNet, rtTable int, log logger.Logger) error {
+	toContainerRule := n.netLink.NewRule()
+	toContainerRule.Dst = containerAddr
+	toContainerRule.Priority = toContainerRulePriority
+	toContainerRule.Table = unix.RT_TABLE_MAIN
+	if err := n.netLink.RuleDel(toContainerRule); err != nil && !containsNoSuchRule(err) {
+		return errors.Wrapf(err, "failed to delete toContainer rule, containerAddr=%s, rtTable=%v", containerAddr.String(), "main")
```

Comment:
`DEL` is a best effort anyway. So, why do we want to return `error` from here? 

https://github.com/containernetworking/cni/blob/main/SPEC.md#del-remove-container-from-network-or-un-apply-modifications

I see, we currently log an error and move on if we fail to delete the route pointing to container IP to container's hostVeth (i.e.,) `toContainerRule` - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/routed-eni-cni-plugin/driver/driver.go#L545

Reply:
The original code merely logs the error if it failed to delete the `toContainerRule`, but it will error out if failed to delete fromContainer rule: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/routed-eni-cni-plugin/driver/driver.go#L556

i normalized this two logic by check for the error condition(!containsNoSuchRule)  to ignore rule not exists error, and return err only if it's other errors.

Reply:
Ok. Since `toContainerRule` is in Main table and pointing to Container's `hostVeth` interface, the rule (ideally) should be purged once we delete the veth pair right? So, even if it fails here for whatever reason the veth deletion operation should take care of it. Am I missing something here? With deletes, since kubelet retries again ...we might be stuck in the delete loop..

`fromContainerRule` on the other hand is pointing to one of the secondary ENIs and the secondary ENI can hold multiple pods and will stay active even after the current pod is deleted.

Reply:
@achevuru i can change it to tolerate the error to be safe to match old code, and log a warning instead only if it's not rule exists error. (i believe there shouldn't be cases trigger that)

Reply:
after more thought, i believe we should still return an error here and let kubelet retry.
1. we already checked for ruleNotExists cases, the rule should be able to be deleted under all other case.
2. if the rule is not deleted, it won't be clean up automatically when veth is delete, only route will.

Reply:
I do agree that apart from rule not being present, it is extremely unlikely to get an error for this operation. If it did fail for some other reason, will retry help here though? If retry doesn't help, my concern is that we will be stuck forever in retry loop. Understand same can be said about `fromContainerrule` but I'm not sure  retrying the operation is going to help here. We can discuss.